### ISSUE TITLE
For temporal weights, Why y should be 3D?

### ISSUE THREAD

**Issue by wddabc** [fchollet/keras#2594]
Hi, if my understanding on the temporal output is correct. It generates a vector rather than a scalar for each example. Considering the number of example, the output should be (nsample, nstep), which should be 2D. In https://github.com/fchollet/keras/blob/master/keras/engine/training.py#L344, the checker seems to enforce y to be 3D, which seems strange to me. Also, the checker, `if sample_weight is not None and len(sample_weight.shape) != 2:` looks very reasonable, as the sample of sample_weight is (nsample, nstep). Shouldn't y have the same shape as sample_weight ?
In my own example:

```
model.compile(loss=_batched_binary_crossentropy, optimizer='adam', sample_weight_mode='temporal')
model.fit(x_train, y_train, sample_weight=w_train, batch_size=_args.batch_size, nb_epoch=_args.nepoch, validation_data=(x_dev, y_dev, w_dev))
```

The shape of x is not important, the shape of y is (nsample, nstep), which means the model predicts a vector with length nstep for each x. w is the weight matrix that have the same shape with y. When I did this. I got an error:

> Timestep-wise sample weighting (use of sample_weight_mode="temporal") is restricted to outputs that are at least 3D, i.e. that have a time dimension.

Is my understanding wrong on the temporal usage? Thanks!
Please make sure that the boxes below are checked before you submit your issue. Thank you!
- [Y] Check that you are up-to-date with the master branch of Keras. You can update with:
  pip install git+git://github.com/fchollet/keras.git --upgrade --no-deps
- [Y] If running on Theano, check that you are up-to-date with the master branch of Theano. You can update with:
  pip install git+git://github.com/Theano/Theano.git --upgrade --no-deps
- [  ] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).
  Basically, it is a conceptual question, not a bug.

**Comment by fchollet** [comment]
I really don't understand what you are asking. Are you asking why masking
time steps in an time series requires to have a time series?

**Comment by wddabc** [comment]
Sorry for the confusion! I must misused the 'temporal' flag I guess.

Let me explain my situation instead. I'm building a model to predict a vector (say length dim) for each example, in which case the batch output is a matrix with shape (nsample, dim). The catch here is each dimension of the vector need to consider a weight when computing the objective. So I also need to pass a weight matrix with shape (nsample, dim), which is the same shape as output. In this case, I need to create a model looks like:
`Model(input=[x, weight], output=y)`

The reason I'm posting this issue is just wondering whether I can pass the weight matrix as sample_weight when fitting the model. It is a cleaner way if I can do that. But currently the weight matrix should only be a 1-D if I don't use 'temporal' flag. In my case, both y and weight are 2-D.

**Comment by fchollet** [comment]
It may be that the solution in your case is to use an output with shape
(nsample, dim, 1) and pass a 2D weight matrix as sample_weight with the
"temporal" flag.

**Comment by esube** [comment]
@fchollet I think, your suggestion wouldn't work. Correct me if I am wrong, but, if you use the 'temporal' flag and you make your 'y' 3D in the way you suggested, it may pass the checking, but, the second dimension of the sample_weight which is dim expects the temporal dimension to be also dim and not 1. 

The best and straight forward way is to allow 2D weight matrices for 2D targets as I seem to repeatedly ask (#2408, etc..) for it and there seems to be another use case here. I am sorry if I seem direct, but, I really want this feature to be in Keras as I grow to like Keras so much.

**Comment by carlthome** [comment]
@esube perhaps you could propose a solution in a pr and move the discussion there?

**Comment by esube** [comment]
@carlthome I am trying this locally. Basically, fixing the placeholder for sample_weight and removing the unnecessary checks in standardize_weights. Other than that, the weighted_objective is already performing the weight multiplication. 

I just need some time to try and test it locally if it works before suggesting a solution. Last time I tried it, it was giving me a dimension mismatch in the objectives and currently I am a bit busy with stuff. But, definitely will get some solution at least locally.

**Comment by azmiara** [comment]
@esube Was this ever resolved??

**Comment by esube** [comment]
Not I know of and I abandoned keras for pytorch for two years now. Pytorch developers appreciate user feedback better and you get better control in your code and you end up with a much shorter, flexible and clean codebase. I suggest you switch to pytorch if you need this feature.
